server/src/main/java/org/elasticsearch/search/aggregations/support/ValuesSourceAggregationBuilder.java

```diff
@@ -144,6 +145,48 @@ public final BucketCardinality bucketCardinality() {
         }
     }
 
+    public abstract static class MetricsAggregationBuilder<VS extends ValuesSource, AB extends ValuesSourceAggregationBuilder<AB>> extends
+        LeafOnly<VS, AB> {
+
+        protected MetricsAggregationBuilder(String name) {
+            super(name);
+        }
+
+        protected MetricsAggregationBuilder(LeafOnly<VS, AB> clone, Builder factoriesBuilder, Map<String, Object> metadata) {
+            super(clone, factoriesBuilder, metadata);
+        }
+
+        protected MetricsAggregationBuilder(StreamInput in) throws IOException {
+            super(in);
+        }
+
+        /** Generated metrics from this aggregation that can be accessed via
+         * {@link org.elasticsearch.search.aggregations.InternalAggregation#getProperty(String)}*/
+        public abstract Set<String> metricNames();
+    }
+
+    public abstract static class SingleMetricAggregationBuilder<VS extends ValuesSource, AB extends ValuesSourceAggregationBuilder<AB>>
+        extends MetricsAggregationBuilder<VS, AB> {
+
+        private static final Set<String> METRIC_NAME = Set.of("value");
```

Comment:
Might be worth a comment that usually the `value` bit isn't required - just pointing at the thing should work.

Reply:
Maybe we should also have a test for that too?

Reply:
I think the situation here is more complicated. 

It is true that if you have an aggregation called `A`, then pointing to the aggregation like `A` or `A.value` is equivalent. But if the aggregation name is `A.B` then they are not. Pointing to such an aggregation as `A.B` ends up in an error as it thinks `B` is a metric. On the other hand `A.B.value` works as expected. Therefore my suggestion is to always use the metric you want to use.

see https://github.com/elastic/elasticsearch/pull/77501.

Reply:
Wow. Sadness. Shows how much I actually use this.